### ISSUE TITLE
3d human pose from tf-pose-estimation

### ISSUE THREAD

**Issue by ArashHosseini** [ildoonet/tf-pose-estimation#384]
@ildoonet  thanks for the great work, there is now the possibility to estimating 3d human poses by using the 2d estimated output from forked [_tf-pose-estimation_](https://github.com/ArashHosseini/tf-pose-estimation) in [3d-pose-basline](https://github.com/ArashHosseini/3d-pose-baseline), check [simple change diff](https://github.com/ArashHosseini/tf-pose-estimation/commit/eb25b197b3c0ed2d424513dbbe2565e910a736d1) for more information
thx

**Comment by highway007** [comment]
@ArashHosseini 
hi, i had changed those .py follow yours, and  there is still no json file. can you tell me what i should do?

**Comment by ArashHosseini** [comment]
@highway007 what is the your command?

**Comment by highway007** [comment]
@ArashHosseini 
thx for reply...my mistake was that i didn't use the absolute path(output_json)
now i got  json file recording 'pose_keypoints_2d' not 'pose_keypoints_3d', like:
{"people": [{"pose_keypoints_2d": [812, 176, 806, 278, 723, 270, 723, 395, 841, 290, 889, 278, 907, 387, 919, 462, 741, 571, 0.0, 0.0, 0.0, 0.0, 812, 575, 0.0, 0.0, 0.0, 0.0, 788, 160, 830, 160, 759, 172, 859, 168]}]}
is there any ways to get the 'pose_keypoints_3d' json file?

**Comment by Sindreedelange** [comment]
Can I piggyback on this issue and ask how you got the 2D .json output, what those values represent, and if they can be 'translated' to "Caffe Openpose's" keypoints .json file?

**Comment by ArashHosseini** [comment]
@Sindreedelange check the changes in [diff](https://github.com/ArashHosseini/tf-pose-estimation/commit/eb25b197b3c0ed2d424513dbbe2565e910a736d1), the output is already matching the Openpose output....you can feed this as input json for 3d-pose-baseline

**Comment by Sindreedelange** [comment]
@ArashHosseini, that is great! 

I get the *_keypoints.json files, however they are empty:
{"people": []}

I believe I have copied everything from your link (and the ones for *run_video.py*) - do you know where I might have misstepped? 

Here is my command
> python run_video.py --video=./videos/jumping_jacks_2.mp4 --resolution=432x368 --model=mobilenet_thin --output_json tmp/

**Comment by ArashHosseini** [comment]
@Sindreedelange Did you also followed this [commit ](
https://github.com/ArashHosseini/tf-pose-estimation/commit/9a8509815081bdc246830bb031429fdf0a651e3f) ?

**Comment by Sindreedelange** [comment]
@ArashHosseini Yes.

**Comment by ArashHosseini** [comment]
@Sindreedelange can you See the joints in the retured Image from TfPoseEstimator.draw_humans func ?

**Comment by Sindreedelange** [comment]
@ArashHosseini No. When --showBG = True it plays the video with no skeleton (and the FPS information), and --showBG = False gives just a black video player, with the FPS information.

**Comment by ArashHosseini** [comment]
@Sindreedelange Can you change line 403 and 411 in drawHumans func to be similar to line 411 # npimg = cv2.line(npimg, centers[pair[0]], centers[pair[1]], common.CocoColors[pair_order], 3) ,lets debug the Image Issue first.....

**Comment by Sindreedelange** [comment]
@ArashHosseini Copied line 411 = *cv2.line(npimg, centers[pair[0]], centers[pair[1]], common.CocoColors[pair_order], 3)* and replaced line 403 = *cv2.circle(npimg, center, 8, common.CocoColors[i], thickness=3, lineType=8, shift=0)* with 411. There is no change in behaviour when running.

**Comment by Sindreedelange** [comment]
@ArashHosseini I believe I have found the problem: 
*humans = e.inference(image)* gives an empty array - I assume this is not intended?

**Comment by ArashHosseini** [comment]
@Sindreedelange can you change the Logger Level to debug in estimator.py....

**Comment by Sindreedelange** [comment]
@ArashHosseini 
> [2019-02-12 17:27:01,835] [TfPoseEstimator] [DEBUG] inference+ original shape=340x256
[2019-02-12 17:27:01,898] [TfPoseEstimator] [DEBUG] inference- heatMat=54x46 pafMat=54x46
[2019-02-12 17:27:01,898] [TfPoseEstimator] [DEBUG] estimate time=0.00013

**Comment by Sindreedelange** [comment]
@ArashHosseini I get (seemingly) correct output when running:
*python run.py --image=./videos/flipped/frame0_flipped.jpg --output_json=tmp/*, which I guess is just restating the point made above from the Notebook, however I find it quite weird that the program is able to find the human in a frame, inputed by a path, but not by an array?

**Comment by ArashHosseini** [comment]
@Sindreedelange, thx can you please make a fresh fork from Upstream or redo the Changes from the commits , then lets run run_video.py so we See the Image included joibts, assume there are People in Image........

**Comment by Sindreedelange** [comment]
@ArashHosseini 
*git reset --hard*, so that the only "new" file is *tf_pose/pafprocess/pafprocess_wrap.cxx*, which I understood from the README that was necessary. 

The program is still __not__ able to recognize humans when running *run_video.py*, but still able when running *run.py*:
![image](https://user-images.githubusercontent.com/9415625/52652941-48813700-2eef-11e9-9473-6598effd8f57.png)

**Comment by ArashHosseini** [comment]
https://github.com/ildoonet/tf-pose-estimation/blob/dbc89c7c7efdf5aaf56b3bb0f53ef32ea3eaff76/run_video.py#L48 , so you See the Image without drawing but with fps Label on it?

**Comment by Sindreedelange** [comment]
@ArashHosseini That is correct.

**Comment by Sindreedelange** [comment]
@ArashHosseini 

Edited line 46 in *run_video.py*:
̀image = TfPoseEstimator.draw_humans(image, humans, imgcopy=False)`
--> 
`humans = e.inference(image, resize_to_default=(w > 0 and h > 0), upsample_size=args.resize_out_ratio)`

and added the necessary parameters.

Now *run_video.py* seems to be working, so it is on to implementing the .json functionality.

**Comment by Sindreedelange** [comment]
@ArashHosseini It seems to be working now:
*run_video.py*
`humans = e.inference(image)` 
--> 
`humans = e.inference(image, resize_to_default=(w > 0 and h > 0), upsample_size=args.resize_out_ratio)` (copied from *run.py*)
and added 
`parser.add_argument('--resize-out-ratio', type=float, default=4.0, help='if provided, resize heatmaps before they are post-processed. default=1.0')`

**Comment by ArashHosseini** [comment]
What is den lenght of pafprocess.get_num_humans() in 
https://github.com/ildoonet/tf-pose-estimation/blob/dbc89c7c7efdf5aaf56b3bb0f53ef32ea3eaff76/tf_pose/estimator.py#L272 ...inference is calling it

**Comment by ArashHosseini** [comment]
@Sindreedelange so its work now? Can you provide a diff...?

**Comment by Sindreedelange** [comment]
Seemingly, at least. Not sure what type of diff is best, but here is something at least:
![image](https://user-images.githubusercontent.com/9415625/52656384-87ff5180-2ef6-11e9-8273-40bc1e73bb9a.png)

**Comment by Sindreedelange** [comment]
@ArashHosseini Seems so. 

Not sure what is the best way of providing a diff, but here is at least something:
![image](https://user-images.githubusercontent.com/9415625/52656439-a8c7a700-2ef6-11e9-93d5-b3d76c008d0a.png)
Please do let me know if there is a better way of doing this.

**Comment by ArashHosseini** [comment]
@Sindreedelange its great, thx.......i guess now you can Start modify and write the frame-keypoints.json file.......

**Comment by Sindreedelange** [comment]
@ArashHosseini Already done, and it works great! I think Openpose normalizes their values, so on to this now. Thank you so much!

**Comment by ArashHosseini** [comment]
@Sindreedelange I'm glad that it works, I wish I could help you more, I'm on a 6 hour train ride, the only thing I had was my smartphone ....

**Comment by Sindreedelange** [comment]
@ArashHosseini No worries, I truly appreciate your effort! Maybe you could explain to me why it works now, when you get to your computer, because I have no clue as to why it works - total shot in the dark.

**Comment by vishnu41** [comment]
@ArashHosseini 
Hey,
when i am trying to pass the json to the real time code by the following command
 python3.5 src/openpose_3dpose_sandbox_realtime.py --camera_frame --residual --batch_norm --dropout 0.5 --max_norm --evaluateActionWise --use_sh --epochs 200 --load 4874200 --pose_estimation_json /home/drive/tf-pose-estimation-master/json

![error](https://user-images.githubusercontent.com/40159472/54339082-cbfe8700-4659-11e9-9ced-9bb46bf8a911.png)
Can you help me here ASAP.
Thanks

**Comment by ArashHosseini** [comment]
Hey @vishnu41, please replace “xy“ by “_data“ in line 86, let it me know if its working, thx

**Comment by vishnu41** [comment]
Yea, i changed
But  not working . same error

**Comment by ArashHosseini** [comment]
@vishnu41, and you Start your Pose-estimator and 3D-baseline on the same Time?

**Comment by vishnu41** [comment]
Yes, i did and i am getting this error
same time in the sense parallely in two terminals no?
![error2](https://user-images.githubusercontent.com/40159472/54344241-682e8b00-4666-11e9-8892-418870a70260.png)

**Comment by ArashHosseini** [comment]
@vishnu41 Yes tf-Pose-estimator is Writing and 3d-baseline is Reading and showing in parallel. The error has Changed.....you know why?

**Comment by vishnu41** [comment]
@ArashHosseini Yea i understood that,But still it shows no image error, what should i do? 
I thought program will automatically saves the image

**Comment by vishnu41** [comment]
@ArashHosseini and i am getting only the tf openpose output in the window and not the 3d baseline

**Comment by vishnu41** [comment]
hey,
Its not woking for me, help me fast , in a hurry.
Reply ASAP
Thanks

**Comment by ArashHosseini** [comment]
@vishnu41,
Is the code writing any png?
Is the plt.show() in line 211 showing something? 
 Is the „not-realtime“ Version working correct with the same json‘s? (the First Frame Should Contain all joints?

**Comment by vishnu41** [comment]
It writes png, but there is nothing in there,
And plt.show() shows images
sometimes its getting terminated automatically.
help me here and i need that interpolation thing also, thats not generating

**Comment by madhuvaani** [comment]
@ArashHosseini i was using your code to get the 3d coordinates , but the coordinates i am getting looks like 24 4d points, Tell me what change i should make to get the 3d points
My goal is to create bvh file , so i need the 3d coordinates, can you tell the steps

**Comment by ArashHosseini** [comment]
@vishnu41, @madhuvanni, i can‘t touch the code until next week Tuesday, hope you forgive me...

**Comment by ArashHosseini** [comment]
@vishnu41 can you please upload one of the json files? thanks

**Comment by madhuvaani** [comment]
@ArashHosseini not an issue

**Comment by vishnu41** [comment]
[000000000000_keypoints.txt](https://github.com/ildoonet/tf-pose-estimation/files/2981590/000000000000_keypoints.txt)

**Comment by ArashHosseini** [comment]
@vishnu41. i have to ask for more, can you please upload the sequence as zip file?,  thanks

**Comment by vishnu41** [comment]
[json.zip](https://github.com/ildoonet/tf-pose-estimation/files/2982791/json.zip)
these are  the jsons only for the upper part , since i placed the camera in such a way.And what is that 4D points? Those are theta? , I am not getting 3d points

**Comment by ArashHosseini** [comment]
@vishnu41, here the upper body output, is the assumption correct that no lower body parts are in the picture?, (you already mentioned it)
![pose_frame_000000000021](https://user-images.githubusercontent.com/9273754/54606573-83c1d900-4a4c-11e9-8328-5956f91846a5.png)

**Comment by ArashHosseini** [comment]
@madhuvaani this question is often asked [#108](https://github.com/una-dinosauria/3d-pose-baseline/issues/108) I'm going to explain it a bit clearer, but it will take some time. thanks

**Comment by ironllamagirl** [comment]
> @highway007 you have to forward the json to [3d-pose-basline ](https://github.com/ArashHosseini/3d-pose-baseline)

What is the command used for this please?

**Comment by ironllamagirl** [comment]
> @madhuvaani this question is often asked [#108](https://github.com/una-dinosauria/3d-pose-baseline/issues/108) I'm going to explain it a bit clearer, but it will take some time. thanks

How do I extract the 3D coordinates in a json file?

**Comment by ironllamagirl** [comment]
> @Sindreedelange check the changes in [diff](https://github.com/ArashHosseini/tf-pose-estimation/commit/eb25b197b3c0ed2d424513dbbe2565e910a736d1), the output is already matching the Openpose output....you can feed this as input json for 3d-pose-baseline

What is this code used for? why multiply the x and y coordinates by the width and heigh and add 0.5? 
 
` body_part = human.body_parts[i]
                center = (int(body_part.x * image_w + 0.5), int(body_part.y * image_h + 0.5))
                centers[i] = center`

**Comment by ironllamagirl** [comment]
> @Sindreedelange check the changes in [diff](https://github.com/ArashHosseini/tf-pose-estimation/commit/eb25b197b3c0ed2d424513dbbe2565e910a736d1), the output is already matching the Openpose output....you can feed this as input json for 3d-pose-baseline

I have followed your code and it seems to work, however I do not understand fully where exactly in the pipeline of the network did we feed the 2D coordinate estimations to the 3d pose baseline? 

3d pose baseline seems to use predictions from stacked hourglass.. it gets the 2d predictions from 3d ground truth data? In this case we will use our own predictions. Where is that intervention happening in the code? 

I used the json files from openpose, but in the terminal I still get log info referring to stacked hourglass and human 3.6M dataset, in h5 files. A little lost and confused in the pipeline here. Any indication or explanation would be much appreciated!

**Comment by igorgbs** [comment]
> @ArashHosseini
> thx for reply...my mistake was that i didn't use the absolute path(output_json)
> now i got json file recording 'pose_keypoints_2d' not 'pose_keypoints_3d', like:
> {"people": [{"pose_keypoints_2d": [812, 176, 806, 278, 723, 270, 723, 395, 841, 290, 889, 278, 907, 387, 919, 462, 741, 571, 0.0, 0.0, 0.0, 0.0, 812, 575, 0.0, 0.0, 0.0, 0.0, 788, 160, 830, 160, 759, 172, 859, 168]}]}
> is there any ways to get the 'pose_keypoints_3d' json file?

where did you find the function to get output from image?

**Comment by ArashHosseini** [comment]
@igorgbs what you mean exactly 

> get output from image?

**Comment by PrashantSaikia** [comment]
> @Sindreedelange check the changes in [diff](https://github.com/ArashHosseini/tf-pose-estimation/commit/eb25b197b3c0ed2d424513dbbe2565e910a736d1), the output is already matching the Openpose output....you can feed this as input json for 3d-pose-baseline

Hi @ArashHosseini, thanks for the commit in your repo where you modified `tf_pose/estimator.py` to give it the option of converting the video to a JSON file. However, when I copied the relevant parts into the function, and tried to save the JSON, it only saved the JSON of the first frame. In the original OpenPose (using Caffe), a JSON is saved for each frame. Do you know how to go about recreating that functionality here? Many thanks in advance!

**Comment by ArashHosseini** [comment]
Hi @PrashantSaikia, 
check this [line](https://github.com/ArashHosseini/tf-pose-estimation/blob/9a8509815081bdc246830bb031429fdf0a651e3f/run_webcam.py#L61) 
```
    frame = 0
    while True:
        ret_val, image = cam.read()

        logger.debug('image process+')
        humans = e.inference(image, resize_to_default=(w > 0 and h > 0), upsample_size=args.resize_out_ratio)

        logger.debug('postprocess+')


        -->     image = TfPoseEstimator.draw_humans(image, humans, imgcopy=False, frame=frame, output_json_dir=args.output_json)
```
inside the while loop, we are calling `tf_pose/estimator.draw_humans()` with the frame and we are increasing the json [here](https://github.com/ArashHosseini/tf-pose-estimation/blob/9a8509815081bdc246830bb031429fdf0a651e3f/tf_pose/estimator.py#L416) based on frame. check if you have all parts after copying

**Comment by PrashantSaikia** [comment]
Hi @ArashHosseini Yes, my bad, I had not copied the `frame += 1` line, oops.

I have another (not related to this) issue that I am trying to resolve. I am not able to access/run the `run_webcam.py` function from outside the `tf_pose_estimation` directory. Any idea on how to get it working from outside this directory? I have detailed the issue [here](https://stackoverflow.com/questions/63573425/how-do-i-run-tf-pose-estimation-from-outside-the-directory). I had faced the same issue with the original openpose as well, where I was not able to get `OpenPoseDemo.exe` running from outside the home directory, which I have detailed [here](https://stackoverflow.com/questions/63171646/how-to-run-open-pose-binary-exe-from-within-a-python-script).

If you could give some suggestions, that would be awesome! Thanks!
